### DISCUSSION TITLE
How to start mitmproxy through python instead of starting mitmproxy through command line.

### DISCUSSION THREAD

**Discussion by Potato-OvO** [mitmproxy/mitmproxy#6039] [Q&A]
I want to control mitmproxy as a python program, instead of starting mitmproxy by calling the command line through python (this method makes me feel that the program is out of control).

My version is as follows:

Mitmproxy: 8.0.0
Python:    3.9.12
OpenSSL:   OpenSSL 1.1.1n  15 Mar 2022
Platform:  Windows-10-10.0.22621-SP0

**Comment by sujaldev** [comment]
Depending on what exactly your goal is here, either see [addons](https://docs.mitmproxy.org/stable/addons-overview/) or take a look at:

https://github.com/mitmproxy/mitmproxy/blob/04d9249ab18cd7bd8b54958714d24614f27863b5/pyproject.toml#L85-L88

you could call these directly after installing mitmproxy:
```
from mitmproxy.tools.main import mitmproxy
from mitmproxy.tools.main import mitmdump
from mitmproxy.tools.main import mitmweb
```

**Reply by Potato-OvO** [reply]
I'm new to mitmproxy and I don't know what you express. But I still appreciate your answer.

My requirement is to control the browser through selenium, then use mitmproxy to capture packets, and save the data processed by mitmproxy as a variable for other py file programs to call.
But when I use the python command to control mitmproxy (subprocess.Popen("mitmdump.exe","--listen-host","127.0.0.1", "--listen-port", f"{port}","-s",f"{program_path}", "--quiet", "--mode",f"upstream:{upstream}"])), I cannot use that saved variable in other py files.

Although I can save it as a csv file and then read the file in other programs. But this feels dumb.

So I wonder if I can call mitmproxy directly through python, similar to this:
```
from mitmproxy import options
from mitmproxy.proxy import config, server
from mitmproxy.tools.dump import DumpMaster

opts = options.Options()
opts.add_option("ssl_insecure", bool, True, "")
handler = DataHandler()
proxy_config = config.ProxyConfig(opts)
proxy_server = server.ProxyServer(proxy_config)
dump_master = DumpMaster(opts)
dump_master.server = proxy_server
dump_master.addons.add(handler)
```
The above are just examples, it is wrong.

I don't know if I express clearly. I'm still very grateful for your answer.

**Reply by sujaldev** [reply]
Let me see if I understand you correctly, you want to run selenium and capture all traffic generated by it and then process it later with another python script?

**Reply by sujaldev** [reply]
By the way, you can use a combination of `-w` and `-r` flags to save the entire stream during a session and read it back again during another.

**Reply by Potato-OvO** [reply]
> Let me see if I understand you correctly, you want to run selenium and capture all traffic generated by it and then process it later with another python script?

Yes, that's what I'm currently doing.

> By the way, you can use a combination of `-w` and `-r` flags to save the entire stream during a session and read it back again during another.



> By the way, you can use a combination of `-w` and `-r` flags to save the entire stream during a session and read it back again during another.

I got you. I just want to be able to operate mitmproxy through python code (just like the wrong sample code I gave you at the beginning). Although my needs can be achieved through -r or -w.

Thanks again for your reply.

**Reply by mwilkowski80** [reply]
If you do not understand pyproject.toml then read about it first, check the syntax and then study these functions as listed by the colleague.

**Comment by y0un9kane** [comment]
Did you get an answer? I am also looking for this method. I have tried the following approaches.

```
  ws_url = "ws://127.0.0.1:8081"  
  ws = websocket.create_connection(ws_url)
```
 
  with open("mitmproxy_messages1.txt", "a") as file:
```
      while True:
          message = ws.recv()
  
          file.write(f"{message}\n")
  
          print(f"Received message from Mitmproxy: {message}")
```
But this doesn't retrieve the content like response.body. I'm still searching for answers by consulting all available resources.

**Comment by KatkaBielcikova** [comment]
This worked for me for python 3.12
https://stackoverflow.com/a/79660673/13312704
